File: survival-texturepacker/src/main/java/net/survival/texturepacker/BitmapComponent.java
package net.survival.texturepacker;

import java.awt.Color;
import java.awt.Graphics;

import javax.swing.JComponent;

import net.survival.graphics.Bitmap;

class BitmapComponent extends JComponent {
    private static final long serialVersionUID = 1L;

    private Bitmap bitmap;

    public BitmapComponent() {}

    public Bitmap getBitmap() {
        return bitmap;
    }

    public void setBitmap(Bitmap bitmap) {
        this.bitmap = bitmap;
        repaint();
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);

        g.setColor(Color.BLACK);
        g.fillRect(0, 0, getWidth(), getHeight());

        if (bitmap != null) {
            var width = Math.min(getWidth(), bitmap.getWidth());
            var height = Math.min(getHeight(), bitmap.getHeight());

            for (var y = 0; y < height; ++y) {
                for (var x = 0; x < width; ++x) {
                    var srcColor = bitmap.getPixel(x, y);
                    var red = srcColor & 0xFF;
                    var green = (srcColor & 0xFF00) >>> 8;
                    var blue = (srcColor & 0xFF0000) >>> 16;

                    var destColor = (red << 16) | (green << 8) | blue;

                    g.setColor(new Color(destColor));
                    g.fillRect(x, y, 1, 1);
                }
            }
        }
    }
}

File: survival-texturepacker/src/main/java/net/survival/texturepacker/BitmapComponent.kt
package net.survival.texturepacker

import net.survival.graphics.Bitmap
import java.awt.Color
import java.awt.Graphics
import javax.swing.JComponent
import kotlin.math.min

internal class BitmapComponent: JComponent() {
    var bmp: Bitmap? = null

    fun getBitmap(): Bitmap? {
        return bmp
    }

    fun setBitmap(bitmap: Bitmap?) {
        this.bmp = bitmap
        repaint()
    }

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        g.color = Color.BLACK
        g.fillRect(0, 0, width, height)

        if (bmp != null) {
            val width = min(width, bmp!!.width)
            val height = min(height, bmp!!.height)
            for (y in 0 until height) {
                for (x in 0 until width) {
                    val srcColor = bmp!!.getPixel(x, y)
                    val red = srcColor and 0xFF
                    val green = srcColor and 0xFF00 ushr 8
                    val blue = srcColor and 0xFF0000 ushr 16
                    val destColor = red shl 16 or (green shl 8) or blue
                    g.color = Color(destColor)
                    g.fillRect(x, y, 1, 1)
                }
            }
        }
    }
}